File: elements/src/main/java/com/rainmore/cms/domains/users/Permission.java
package com.rainmore.cms.domains.users;

import com.rainmore.cms.domains.Auditable;
import com.rainmore.cms.domains.HasId;
import com.rainmore.cms.domains.Nameable;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;

@Entity(name = "users.permission")
@Table(name = "userPermissions")
public class Permission extends Auditable implements HasId<Long>, Nameable {

    private Long id;
    private String name;
    private Permission parent;

    @Id
    @Column(nullable = false)
    @GeneratedValue
    @Override
    public Long getId() {
        return id;
    }

    @Override
    public void setId(Long id) {
        this.id = id;
    }

    @Column(nullable = false)
    @Override
    @NotEmpty
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @ManyToOne
    @JoinColumn(name = "parentId")
    public Permission getParent() {
        return parent;
    }

    public void setParent(Permission parent) {
        this.parent = parent;
    }
}

File: elements/src/main/kotlin/com/rainmore/cms/domains/core/users/Permission.kt
package com.rainmore.cms.domains.core.users

import com.rainmore.cms.domains.HasId
import com.rainmore.cms.domains.Nameable
import com.rainmore.cms.domains.core.ArchivableAuditableDomain
import javax.persistence.*

@Entity(name = "core.users.permission")
@Table(name = "userPermissions")
data class Permission
(
        @Id
        @GeneratedValue(strategy = GenerationType.AUTO)
        override var id: Long? = null,

        @Column(nullable = false)
        override var name: String? = null
) : ArchivableAuditableDomain<Permission>(), Nameable, HasId<Long> {

    @ManyToOne
    @JoinColumn(name = "parentId")
    var parent: Permission? = null

}
